From 29586df8806286299cc4f8378c632cb7d0c55adf Mon Sep 17 00:00:00 2001
From: Leonid Plyushch <leonid.plyushch@gmail.com>
Date: Thu, 11 May 2017 19:17:18 +0300
Subject: [PATCH] failsafe mode: allow bypassing password login only if file
 '/sdcard/.termux_bypass_login' was found

NOTE: password login protection is used only in special rootfs
build (https://github.com/xeffyr/Termux-RootFS). It is needed when
Termux app is pinned (via Android Screen Pinning).
---
 app/src/main/java/com/termux/app/TermuxActivity.java | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/app/src/main/java/com/termux/app/TermuxActivity.java b/app/src/main/java/com/termux/app/TermuxActivity.java
index eac48ab..31167e2 100644
--- a/app/src/main/java/com/termux/app/TermuxActivity.java
+++ b/app/src/main/java/com/termux/app/TermuxActivity.java
@@ -25,6 +25,7 @@ import android.media.SoundPool;
 import android.net.Uri;
 import android.os.Build;
 import android.os.Bundle;
+import android.os.Environment;
 import android.os.IBinder;
 import android.os.Vibrator;
 import android.support.annotation.NonNull;
@@ -609,6 +610,19 @@ public final class TermuxActivity extends Activity implements ServiceConnection
             new AlertDialog.Builder(this).setTitle(R.string.max_terminals_reached_title).setMessage(R.string.max_terminals_reached_message)
                 .setPositiveButton(android.R.string.ok, null).show();
         } else {
+            File bypassLoginProtectionFile = new File(Environment.getExternalStorageDirectory() + "/.termux_bypass_login");
+            File termuxLoginDataFile = new File("/data/data/com.termux/files/usr/etc/login.pwd");
+
+            if(failSafe && termuxLoginDataFile.exists() && !bypassLoginProtectionFile.exists()) {
+                StringBuilder sb = new StringBuilder();
+                sb.append("You need to create file\n'");
+                sb.append(bypassLoginProtectionFile.getPath());
+                sb.append("'\nfor using failsafe mode.");
+
+                showToast(sb.toString(), true);
+                return;
+            }
+
             String executablePath = (failSafe ? "/system/bin/sh" : null);
             TerminalSession newSession = mTermService.createTermSession(executablePath, null, null, failSafe);
             if (sessionName != null) {
-- 
2.13.0

